pennylane/templates/state_preparations.py

```diff
@@ -0,0 +1,315 @@
+# Copyright 2018 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+r"""
+State Preparations
+==================
+
+**Module name:** :mod:`pennylane.templates.state_preparations`
+
+.. currentmodule:: pennylane.templates.state_preparations
+
+This module provides routines that prepare a given state using only
+elementary gates.
+
+Qubit architectures
+-------------------
+
+.. autosummary::
+
+    BasisStatePreparation
+    MottonenStatePreparation
+
+Code details
+^^^^^^^^^^^^
+"""
+import math
+from collections.abc import Iterable
+
+import numpy as np
+import pennylane as qml
+from scipy import sparse
+from sympy.combinatorics.graycode import GrayCode
+
+# pylint: disable=len-as-condition
+
+def BasisStatePreparation(basis_state, wires):
+    r"""
+    Prepares a basis state on the given wires using a sequence of Pauli X gates.
+
+    Args:
+        basis_state (array): Input array of shape ``(N,)``, where N is the number of wires
+            the state preparation acts on. ``N`` must be smaller or equal to the total
+            number of wires.
+        wires (Sequence[int]): sequence of qubit indices that the template acts on
+    """
+
+    if not isinstance(wires, Iterable):
+        raise ValueError("Wires must be passed as a list of integers; got {}.".format(wires))
+
+    if not len(basis_state) == len(wires):
+        raise ValueError(
+            "Number of qubits must be equal to the number of wires, which is {}; "
+            "got {}.".format(len(wires), len(basis_state))
+        )
+
+    if any([x not in [0, 1] for x in basis_state]):
+        raise ValueError("Basis state must only consist of 0s and 1s, got {}".format(basis_state))
+
+    for wire, state in zip(wires, basis_state):
+        if state == 1:
+            qml.PauliX(wire)
+
+
+def _matrix_M_entry(row, col):
+    """The matrix entry for the angle computation.
+
+    Args:
+        row (int): one-based row number
+        col (int): one-based column number
+
+    Returns:
+        (float): matrix entry
+    """
+    # (col >> 1) ^ col is the Gray code of col
+    b_and_g = row & ((col >> 1) ^ col)
+    sum_of_ones = 0
+    while b_and_g > 0:
+        if b_and_g & 0b1:
+            sum_of_ones += 1
+
+        b_and_g = b_and_g >> 1
+
+    return (-1) ** sum_of_ones
```

Comment:
This function is very hard to understand 😆 Would be good to add comments explaining the process.

Reply:
Basically it implements the construction of M from the paper, I just copied it from Carsten Blank and I simply can't explain it :sweat_smile:

Reply:
Glanced into the paper a bit and it seems that matrix M, whose entries this function builds up, is responsible for the mapping between the angles inputted to the uniformly controlled rotation (``thetas``) and the angles of the single qubit rotations (``alphas``). This is the mapping that yields the new parameters when the uniformly controlled rotation is decomposed into single qubit rotations and CNOTs.

In [their previous paper] it is stated: "The position of the control node in the *lth* CNOT gate is set to
match the position where the *l<sup>th</sup>* and *(l + 1)<sup>th</sup>* bit strings *g<sub>l−1</sub>* and *g<sub>l</sub>* of the binary reflected Gray code [18] differ.

So eventually these positions are computed here for each of the CNOTs and stored in the M matrix.

[their previous paper]: https://arxiv.org/abs/quant-ph/0404089